File: src/disco.c
#include "disco.h"
#include "stanza.h"
#include "config.h"
#include "util.h"
#include <libxml/tree.h>

void disco_handle_info(session_t *s, xmlNodePtr stanza) {
    xmlChar *id = xmlGetProp(stanza, (const xmlChar *)"id");

    char full_jid[768];
    jid_full(s->jid_local, s->jid_domain, s->jid_resource,
             full_jid, sizeof(full_jid));

    xmlNodePtr result = xmlNewNode(NULL, (const xmlChar *)"iq");
    xmlNewProp(result, (const xmlChar *)"type", (const xmlChar *)"result");
    xmlNewProp(result, (const xmlChar *)"from", (const xmlChar *)g_config.domain);
    xmlNewProp(result, (const xmlChar *)"to", (const xmlChar *)full_jid);
    if (id) {
        xmlNewProp(result, (const xmlChar *)"id", id);
        xmlFree(id);
    }

    xmlNodePtr query = xmlNewChild(result, NULL, (const xmlChar *)"query", NULL);
    xmlNsPtr ns = xmlNewNs(query,
        (const xmlChar *)"http://jabber.org/protocol/disco#info", NULL);
    xmlSetNs(query, ns);

    /* Identity */
    xmlNodePtr identity = xmlNewChild(query, ns, (const xmlChar *)"identity", NULL);
    xmlNewProp(identity, (const xmlChar *)"category", (const xmlChar *)"server");
    xmlNewProp(identity, (const xmlChar *)"type", (const xmlChar *)"im");
    xmlNewProp(identity, (const xmlChar *)"name", (const xmlChar *)"xmppd");

    /* Features */
    const char *features[] = {
        "http://jabber.org/protocol/disco#info",
        "http://jabber.org/protocol/disco#items",
        "jabber:iq:roster",
        "urn:xmpp:delay",
        NULL
    };

    for (int i = 0; features[i]; i++) {
        xmlNodePtr feat = xmlNewChild(query, ns, (const xmlChar *)"feature", NULL);
        xmlNewProp(feat, (const xmlChar *)"var", (const xmlChar *)features[i]);
    }

    stanza_send(s, result);
    xmlFreeNode(result);
}

void disco_handle_items(session_t *s, xmlNodePtr stanza) {
    xmlChar *id = xmlGetProp(stanza, (const xmlChar *)"id");

    char full_jid[768];
    jid_full(s->jid_local, s->jid_domain, s->jid_resource,
             full_jid, sizeof(full_jid));

    xmlNodePtr result = xmlNewNode(NULL, (const xmlChar *)"iq");
    xmlNewProp(result, (const xmlChar *)"type", (const xmlChar *)"result");
    xmlNewProp(result, (const xmlChar *)"from", (const xmlChar *)g_config.domain);
    xmlNewProp(result, (const xmlChar *)"to", (const xmlChar *)full_jid);
    if (id) {
        xmlNewProp(result, (const xmlChar *)"id", id);
        xmlFree(id);
    }

    xmlNodePtr query = xmlNewChild(result, NULL, (const xmlChar *)"query", NULL);
    xmlNsPtr ns = xmlNewNs(query,
        (const xmlChar *)"http://jabber.org/protocol/disco#items", NULL);
    xmlSetNs(query, ns);

    /* Empty items list */

    stanza_send(s, result);
    xmlFreeNode(result);
}

File: internal/disco/disco.go
package disco

import (
	"fmt"
	"jabber/internal/session"
	"jabber/internal/xml"
	"strings"
)

// HandleInfo responds to disco#info queries.
func HandleInfo(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	id := node.Attr("id")
	fullJID := s.FullJID()

	var b strings.Builder
	fmt.Fprintf(&b, "<iq type='result' from='%s' to='%s'", reg.Domain(), escAttr(fullJID))
	if id != "" {
		fmt.Fprintf(&b, " id='%s'", escAttr(id))
	}
	b.WriteString("><query xmlns='http://jabber.org/protocol/disco#info'>")
	b.WriteString("<identity category='server' type='im' name='xmppd'/>")

	features := []string{
		"http://jabber.org/protocol/disco#info",
		"http://jabber.org/protocol/disco#items",
		"jabber:iq:roster",
		"jabber:iq:register",
		"urn:xmpp:delay",
	}
	for _, f := range features {
		fmt.Fprintf(&b, "<feature var='%s'/>", escAttr(f))
	}
	b.WriteString("</query></iq>")
	s.Send(b.String())
	return session.RouteOK
}

// HandleItems responds to disco#items queries (empty list).
func HandleItems(reg session.Registry, s *session.Session, node *xml.XMLNode) session.RouteResult {
	id := node.Attr("id")
	fullJID := s.FullJID()

	var b strings.Builder
	fmt.Fprintf(&b, "<iq type='result' from='%s' to='%s'", reg.Domain(), escAttr(fullJID))
	if id != "" {
		fmt.Fprintf(&b, " id='%s'", escAttr(id))
	}
	b.WriteString("><query xmlns='http://jabber.org/protocol/disco#items'/></iq>")
	s.Send(b.String())
	return session.RouteOK
}

func escAttr(s string) string {
	s = strings.ReplaceAll(s, "&", "&amp;")
	s = strings.ReplaceAll(s, "'", "&apos;")
	s = strings.ReplaceAll(s, "<", "&lt;")
	return s
}
